refac: change `Checkbox` to be a wrapper around `KCheckbox` instead of `VCheckbox`

EshaanAgg opened the pull request:
## Summary
### Description of the change(s) you made
- Changes the `Checkbox` component to make use of the `KCheckbox` internally and adds support for using `v-model` to the same
- The `v-model` is designed to handle three kinds of values: `arrays`, `number`, and `boolean`

### Manual verification steps performed
Ran `studio` after making all the changes and verified manually if the checkboxes were behaving as expected. 

### Does this introduce any tech-debt items?
No, but this PR must be manually tested before it can be merged. This is because all the occurrences of the `Checkbox` right now have props and values recognized by `VCheckbox`. These props need to be added to our new implementation of `Checkbox` as well or need to be handled in the components using them separately. 

## Reviewer guidance
### How can a reviewer test these changes?
Running the application locally.

### Are there any risky areas that deserve extra testing?
All checkbox-related menus and forms.

### Contributor's Checklist
PR process:

- [ ] If this is an important user-facing change, PR or related issue the `CHANGELOG` label been added to this PR. Note: items with this label will be added to the [CHANGELOG](https://github.com/learningequality/studio/blob/master/CHANGELOG.md) at a later time
- [ ] If this includes an internal dependency change, a link to the diff is provided
- [ ] The `docs` label has been added if this introduces a change that needs to be updated in the [user docs](https://kolibri-studio.readthedocs.io/en/latest/index.html)?
- [ ] If any Python requirements have changed, the updated `requirements.txt` files also included in this PR
- [ ] Opportunities for using Google Analytics here are noted
- [ ] Migrations are [safe for a large db](https://www.braintreepayments.com/blog/safe-operations-for-high-volume-postgresql/)

Studio-specifc:

- [ ] All user-facing strings are translated properly
- [ ] The `notranslate` class been added to elements that shouldn't be translated by Google Chrome's automatic translation feature (e.g. icons, user-generated text)
- [ ] All UI components are LTR and RTL compliant
- [ ] Views are organized into `pages`, `components`, and `layouts` directories [as described in the docs](https://github.com/learningequality/studio/blob/vue-refactor/docs/architecture.md#where-does-the-frontend-code-live)
- [ ] Users' storage used is recalculated properly on any changes to main tree files
- [ ] If there new ways this uses user data that needs to be factored into our [Privacy Policy](https://github.com/learningequality/studio/tree/master/contentcuration/contentcuration/templates/policies/text), it has been noted.


Testing:

- [ ] Code is clean and well-commented
- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Any new interactions have been added to the [QA Sheet](https://docs.google.com/spreadsheets/d/1HF4Gy6rb_BLbZoNkZEWZonKFBqPyVEiQq4Ve6XgIYmQ/edit#gid=0)
- [ ] Critical and brittle code paths are covered by unit tests
___

### Reviewer's Checklist
#### This section is for reviewers to fill out.

- [ ] Automated test coverage is satisfactory
- [ ] PR is fully functional
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependency files were updated if necessary (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Contributor is in AUTHORS.md

MisRob commented:
Thanks @EshaanAgg!

For reference for reviewers, this PR is not coming out of the blue :) @EshaanAgg has discussed high-level approach with @bjester and me (https://learningequality.slack.com/archives/C03S2EN192A/p1709751447068349)

bjester commented:
@EshaanAgg Thanks for working on this! Seems there are some tests failing. They might need updated, but I think there's a possibility the `Checkbox` abstraction may need to handle objects too, i.e. setting a boolean for a key in an object.

Here's the summary of the failing tests:
https://github.com/learningequality/studio/actions/runs/8208392765/job/22577441251?pr=4472#step:6:12869

You can run the tests locally with `yarn run test-jest`

EshaanAgg commented:
Thanks a lot for the help with debugging, @bjester! I'll try to make the relevant changes so that the same works! Thanks.

EshaanAgg commented:
I have worked on fixing most of the tets. When running the tests locally, only 2 tests fail, and I am a bit confused as how to I might go about fixing the same:-
1. [Answers Editor](https://github.com/learningequality/studio/blob/unstable/contentcuration/contentcuration/frontend/channelEdit/components/AnswersEditor/AnswersEditor.vue)

LianaHarris360 commented:
For the issue with the Answers Editor component, the state prop is set to false when it should be an array. Updating line 50 in [AnswersEditor.vue](https://github.com/learningequality/studio/blob/5c0d00a1f39557003a5e4ba7b0fadcce0d42ef8b/contentcuration/contentcuration/frontend/channelEdit/components/AnswersEditor/AnswersEditor.vue#L50) to use `v-model` instead of :`input-value` will properly connect the array of correct answer indices to the `state` prop. The other connected issue I noticed is that deselecting and selecting answers does not update the parent component of `AnswersEditor` correctly, if the question can have multiple answers.

EshaanAgg commented:
> For the issue with the Answers Editor component, the state prop is set to false when it should be an array. Updating line 50 in [AnswersEditor.vue](https://github.com/learningequality/studio/blob/5c0d00a1f39557003a5e4ba7b0fadcce0d42ef8b/contentcuration/contentcuration/frontend/channelEdit/components/AnswersEditor/AnswersEditor.vue#L50) to use `v-model` instead of :`input-value` will properly connect the array of correct answer indices to the `state` prop. The other connected issue I noticed is that deselecting and selecting answers does not update the parent component of `AnswersEditor` correctly, if the question can have multiple answers.

Got it. It seem's to be that some logic surrounding all of these interactions and changes needs to be rewritten a bit. Will try to do so ASAP. Thanks so much for your time and help! I really appreciate it!

akolson reviewed (changes requested):
Hi @EshaanAgg. I think the changes made look great. Thanks.

Concerning the failing tests, I think that changing [this](https://github.com/learningequality/studio/blob/4c937ea2c7c61b424589e57cf8125cb9b13a2d44/contentcuration/contentcuration/frontend/shared/views/form/__tests__/contentDefaults.spec.js#L68) to the correct prop should fix the problem for tests failing in `contentDefaults.spec.js`. Hopefully this helps.

EshaanAgg commented:
@akolson Thanks for the help. I have updated the test suite in the `ContentDefaults` component to make use of the suggestion you provided, and it passes all the test suite now! I tried tackling the other two components again today, but am struggling with the implementation and figuring out a way to get the desired reactivity with the way the `getter-setter` based `v-models`. Will try taking out some more time to fix it.

akolson commented:
Hi @EshaanAgg!

I think for your implementation of the `Checkbox` component to be compatible with the current code, It should also accept the `input-value` (or `inputValue`) prop. [`input-value`](https://v2.vuetifyjs.com/en/api/v-checkbox/#props) is a Vuetify(v2) inspired prop that according to the documentation is _The v-model bound value_ of the checkbox. I believe this value is synonymous to the the `checked` prop in the [`KCheckbox`](https://design-system.learningequality.org/kcheckbox/) component in KDS. The difference is that `checked` holds the truthy state value of the checkbox, while `input-value` holds the checkbox's state value which could be for example an `Array`, `boolean`,  or `object` (there could be others, didn't do a lot of auditing). We could say the the `input-value` is `v-model`esque in a way, so I think setting the `input-value` of the `Checkbox` should also set the `v-model` prop of the `KCheckbox`.

Having the `Checkbox` accept both `v-model` and `input-value` is a way for us to keep backward compatibility until we are able to dedicate time to cleanup the usage variations.  So unless the scope of this pr involves the cleanup, I think it would make more sense to have the component backward compatible. @bjester any varying thoughts or corrections?

EshaanAgg commented:
Thanks @akolson @bjester. I am taking a few days off due to some health reasons, but will definitely update the PR as soon as possible. Thanks again for taking out the time to leave such detailed comments and for all the help! I really appreciate the same.

EshaanAgg commented:
Hi! Thanks for all your patience! I have updated the PR such that the test suite passes completely locally and addressed all the linting issues. Hopefully, this PR is ready for review. Thanks again.

akolson commented:
HI @EshaanAgg! Great work on resolving the failing tests. However, it looks like we have one more failing test in `contentcuration/contentcuration/frontend/accounts/pages/__tests__/create.spec.js` I suspect that its related to the `rules` and `hide-details` props that have been removed from the `Checkbox.vue` component and yet used in the `Create.vue` component.

```vue
......
<Checkbox
  v-model="acceptedAgreement"
  :label="$tr('agreement')"
  required
  :rules="tosAndPolicyRules"
  :hide-details="false"
  class="my-1 policy-checkbox"
/>
......
```

I am unsure about the decision that was taken on these props apart from the discussion [here](https://learningequality.slack.com/archives/C03S2EN192A/p1709929294742249?thread_ts=1709751447.068349&cid=C03S2EN192A), so some clarification could help resolve the issue. @bjester?

EshaanAgg commented:
Thanks @akolson. I guess that the best way to handle the same would be to provide support for these props in the `Checkbox`  component itself, as these give seemingly nice-to-have functionality that we may want to use in other UIs across Studio in the future (as these functionalities aren't specific to the `Create` component). If the same sounds good, I can work on implementing them today!

Thoughts @akolson @bjester?

EshaanAgg commented:
Hey! I had completed the addition of support for the `hideDetails` and `rules` prop to our Checkbox component when I realized that we are making use of the `VForm` component's internal method `.validate()` while submitting the same. 
https://github.com/learningequality/studio/blob/fb765d24ad2527e802d98b4a1e52d4f01807d6da/contentcuration/contentcuration/frontend/accounts/pages/Create.vue#L427

Since this method relies on all the children components to be Vuetify components as well, I guess using `KCheckbox` here won't be feasible (unless we refactor the whole page). Just IMO, the best approach right now we can take for this page is to make use of the `VCheckbox` component here directly instead of the `Checkbox` wrapper, so that we don't have to go into full fledge refactoring, and we can continue the migration as well.

akolson commented:
Thanks @EshaanAgg for raising your observations and for your dedication to the issue. We'll create a follow up issue to resolve the failure in `contentcuration/contentcuration/frontend/accounts/pages/__tests__/create.spec.js`. 

We'll approve and merge this issue for now.

akolson reviewed (approved):
Approved!

akolson commented on contentcuration/contentcuration/frontend/accounts/pages/Create.vue:
```diff
@@ -132,7 +132,7 @@
 
         <!-- Agreements -->
-        <Checkbox
+        <VCheckbox
```
This should be changed to `Checkbox` in https://github.com/learningequality/studio/issues/4492

bjester commented:
@EshaanAgg Thank you for your contributions!
